File: iosApp/iosApp/Modules/MnemonicImport/MnemonicImportWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

protocol MnemonicImportWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: MnemonicImportWireframeContext
    ) -> MnemonicImportWireframe
}

final class DefaultMnemonicImportWireframeFactory {
    private let signerStoreService: SignerStoreService
    private let mnemonicService: MnemonicService
    private let passwordService: PasswordService
    private let settingsService: SettingsService
    private let addressService: AddressService

    init(
        signerStoreService: SignerStoreService,
        mnemonicService: MnemonicService,
        passwordService: PasswordService,
        settingsService: SettingsService,
        addressService: AddressService
    ) {
        self.signerStoreService = signerStoreService
        self.mnemonicService = mnemonicService
        self.passwordService = passwordService
        self.settingsService = settingsService
        self.addressService = addressService
    }
}

extension DefaultMnemonicImportWireframeFactory: MnemonicImportWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: MnemonicImportWireframeContext
    ) -> MnemonicImportWireframe {
        DefaultMnemonicImportWireframe(
            parent,
            context: context,
            signerStoreService: signerStoreService,
            mnemonicService: mnemonicService,
            passwordService: passwordService,
            settingsService: settingsService,
            addressService: addressService
        )
    }
}

final class MnemonicImportWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> MnemonicImportWireframeFactory in
            DefaultMnemonicImportWireframeFactory(
                signerStoreService: resolver.resolve(),
                mnemonicService: resolver.resolve(),
                passwordService: resolver.resolve(),
                settingsService: resolver.resolve(),
                addressService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/mnemonicimport/MnemonicImportWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicimport

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.mnemonicImport.MnemonicImportWireframe
import com.sonsofcrypto.web3walletcore.modules.mnemonicImport.MnemonicImportWireframeContext
import com.sonsofcrypto.web3walletcore.services.mnemonic.MnemonicService
import com.sonsofcrypto.web3walletcore.services.password.PasswordService
import smartadapter.internal.extension.name

interface MnemonicImportWireframeFactory {
    fun make(parent: Fragment?, context: MnemonicImportWireframeContext): MnemonicImportWireframe
}

class DefaultMnemonicImportWireframeFactory(
    private val keyStoreService: KeyStoreService,
    private val mnemonicService: MnemonicService,
    private val passwordService: PasswordService,
): MnemonicImportWireframeFactory {

    override fun make(
        parent: Fragment?, context: MnemonicImportWireframeContext
    ): MnemonicImportWireframe = DefaultMnemonicImportWireframe(
        parent?.let { WeakRef(it) },
        context,
        keyStoreService,
        mnemonicService,
        passwordService,
    )
}

class MnemonicImportWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(MnemonicImportWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultMnemonicImportWireframeFactory(
                it.resolve(KeyStoreService::class.name),
                it.resolve(MnemonicService::class.name),
                it.resolve(PasswordService::class.name),
            )
        }
    }
}